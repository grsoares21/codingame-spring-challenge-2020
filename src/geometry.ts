export interface Point {
  x: number;
  y: number;
}

export function manhattanDistance(pointA: Point, pointB: Point): number {
  return Math.abs(pointA.x - pointB.x) + Math.abs(pointA.y - pointB.y);
}
