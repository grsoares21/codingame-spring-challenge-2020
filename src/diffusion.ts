import { Point } from "./geometry";
import { PriorityQueue } from "./queues";

export function getSignalMatrix(
  map: string[][],
  singalPoint: Point,
  maxSignalDistance: number,
  diffusionFactor: number,
  signalPointValue: number
): number[][] {
  type VisitingPoint = {
    point: Point;
    value: number;
    distanceToSrc: number;
    visited: boolean;
  };

  const pointsQueue = new PriorityQueue<VisitingPoint>({
    comparator: function (pointA, pointB) {
      const pointATotalDistance = pointA.distanceToSrc;
      const pointBTotalDistance = pointB.distanceToSrc;
      return pointATotalDistance - pointBTotalDistance;
    },
  });

  let visitedPointsMatrix: VisitingPoint[][] = [];

  for (let y = 0; y < map.length; y++) {
    visitedPointsMatrix[y] = [];
    for (let x = 0; x < map[y].length; x++) {
      visitedPointsMatrix[y][x] =
        map[y][x] === "#"
          ? {
              point: { x, y },
              value: Number.NEGATIVE_INFINITY,
              distanceToSrc: -1,
              visited: true,
            }
          : {
              point: { x, y },
              distanceToSrc: Number.POSITIVE_INFINITY,
              visited: false,
              value: 0,
            };
    }
  }

  visitedPointsMatrix[singalPoint.y][singalPoint.x].distanceToSrc = 0;
  visitedPointsMatrix[singalPoint.y][singalPoint.x].value = signalPointValue;
  pointsQueue.queue(visitedPointsMatrix[singalPoint.y][singalPoint.x]);

  while (pointsQueue.length > 0) {
    const currentPointToVisit = pointsQueue.dequeue();
    currentPointToVisit.visited = true;

    const width = visitedPointsMatrix[currentPointToVisit.point.y].length;

    const northPoint =
      visitedPointsMatrix[currentPointToVisit.point.y - 1] &&
      visitedPointsMatrix[currentPointToVisit.point.y - 1][
        currentPointToVisit.point.x
      ];
    const southPoint =
      visitedPointsMatrix[currentPointToVisit.point.y + 1] &&
      visitedPointsMatrix[currentPointToVisit.point.y + 1][
        currentPointToVisit.point.x
      ];
    const eastPoint =
      visitedPointsMatrix[currentPointToVisit.point.y][
        (currentPointToVisit.point.x + 1) % width
      ];
    const westPoint =
      visitedPointsMatrix[currentPointToVisit.point.y][
        currentPointToVisit.point.x -
          1 +
          (currentPointToVisit.point.x - 1 < 0 ? width : 0)
      ];
    if (northPoint) {
      if (
        !northPoint.visited &&
        currentPointToVisit.distanceToSrc + 1 < maxSignalDistance
      ) {
        northPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        northPoint.value =
          signalPointValue /
          Math.pow(northPoint.distanceToSrc + 1, diffusionFactor);

        pointsQueue.queue(northPoint);
      }
    }
    if (southPoint) {
      if (
        !southPoint.visited &&
        currentPointToVisit.distanceToSrc + 1 < maxSignalDistance
      ) {
        southPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        southPoint.value =
          signalPointValue /
          Math.pow(southPoint.distanceToSrc + 1, diffusionFactor);

        pointsQueue.queue(southPoint);
      }
    }
    if (eastPoint) {
      if (
        !eastPoint.visited &&
        currentPointToVisit.distanceToSrc + 1 < maxSignalDistance
      ) {
        eastPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        eastPoint.value =
          signalPointValue /
          Math.pow(eastPoint.distanceToSrc + 1, diffusionFactor);

        pointsQueue.queue(eastPoint);
      }
    }
    if (westPoint) {
      if (
        !westPoint.visited &&
        currentPointToVisit.distanceToSrc + 1 < maxSignalDistance
      ) {
        westPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        westPoint.value =
          signalPointValue /
          Math.pow(westPoint.distanceToSrc + 1, diffusionFactor);

        pointsQueue.queue(westPoint);
      }
    }
  }
  let returnedMatrix = [];

  for (let y = 0; y < visitedPointsMatrix.length; ++y) {
    returnedMatrix[y] = [];
    for (let x = 0; x < visitedPointsMatrix[y].length; ++x) {
      returnedMatrix[y][x] = visitedPointsMatrix[y][x].value;
    }
  }
  return visitedPointsMatrix.map((line) =>
    line.map((visitingPoint) => visitingPoint.value)
  );
}
