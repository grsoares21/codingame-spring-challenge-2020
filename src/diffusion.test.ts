import { findPath } from "./geometry";
import { getSignalMatrix } from "./diffusion";

it(`should emit 10 signal with 10 max distance and 0.9 diffusion factor`, () => {
  let map = getMap();
  expect(
    prettyPrintNumberMatrix(getSignalMatrix(map, { x: 15, y: 8 }, 10, 0.9, 9))
  ).toMatchSnapshot();
});

it(`should cross map when emitting signal`, () => {
  let map = getMap();
  expect(
    prettyPrintNumberMatrix(getSignalMatrix(map, { x: 3, y: 4 }, 10, 0.9, 9))
  ).toMatchSnapshot();
  expect(
    prettyPrintNumberMatrix(getSignalMatrix(map, { x: 31, y: 5 }, 10, 0.9, 9))
  ).toMatchSnapshot();
});

it(`should emit 1 signal with 5 max distance and 0.5 diffusion factor`, () => {
  let map = getMap();
  expect(
    prettyPrintNumberMatrix(getSignalMatrix(map, { x: 1, y: 1 }, 5, 0.5, 1))
  ).toMatchSnapshot();
});

it(`should emit -5 signal with 5 max distance and 0.5 diffusion factor`, () => {
  let map = getMap();
  expect(
    prettyPrintNumberMatrix(getSignalMatrix(map, { x: 1, y: 1 }, 5, 0.5, -5))
  ).toMatchSnapshot();
});

function prettyPrintNumberMatrix(matrix: number[][]): string {
  return matrix
    .map((line) =>
      line
        .map((num) => num.toFixed(1))
        .map((num) => (num === "-Infinity" ? "###" : num))
        .join(" ")
    )
    .join("\n");
}

function getMap(): string[][] {
  // prettier-ignore
  return [
    ['#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#'],
    ['#',' ','#',' ',' ',' ','#',' ',' ',' ','#',' ',' ',' ','#','#','#','#','#',' ',' ',' ','#',' ',' ',' ','#',' ',' ',' ','#',' ','#'],
    ['#',' ','#',' ','#','#','#','#','#',' ','#',' ','#','#','#','#','#','#','#','#','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#'],
    ['#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#'],
    ['#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#','#','#','#','#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#'],
    [' ',' ',' ',' ','#',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ','#',' ',' ',' ',' '],
    ['#','#','#',' ','#',' ','#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#'],
    ['#',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ','#'],
    ['#',' ','#','#','#',' ','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#',' ','#','#','#',' ','#'],
    ['#',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ','#'],
    ['#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#'],
    ['#',' ',' ',' ','#',' ',' ',' ',' ',' ','#',' ',' ',' ','#',' ','#',' ','#',' ',' ',' ','#',' ',' ',' ',' ',' ','#',' ',' ',' ','#'],
    ['#',' ','#',' ','#','#','#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#'],
    ['#',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ','#'],
    ['#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#']
  ];
}
