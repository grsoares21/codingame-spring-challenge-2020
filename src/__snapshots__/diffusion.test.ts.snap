// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should cross map when emitting signal 1`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 1.8 ### 2.6 2.1 1.8 ### 0.0 0.0 1.1 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 2.1 ### 3.3 ### ### ### ### ### 1.2 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 2.6 3.3 4.8 3.3 2.6 2.1 1.8 1.6 1.4 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 9.0 ### 2.1 ### 1.6 ### 1.2 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 1.1 ### ### ###
2.1 2.6 3.3 4.8 ### 1.8 ### 1.4 1.2 1.1 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 1.2 1.4 1.6 1.8
### ### ### 3.3 ### 1.6 ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ### 1.1 ### ### ###
### 1.8 2.1 2.6 ### 1.4 1.2 1.1 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 1.6 ### ### ### 1.2 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 1.4 1.2 1.1 0.0 1.1 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should cross map when emitting signal 2`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 0.0 ### 1.1 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 1.2 1.4 1.6 ### 1.2 ###
### 0.0 ### 1.2 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 1.8 ### 1.4 ###
### 1.1 1.2 1.4 1.2 1.1 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 1.1 1.2 1.4 1.6 1.8 2.1 1.8 1.6 ###
### ### ### 1.6 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### 1.1 ### 1.4 ### 2.6 ### ### ###
3.3 2.6 2.1 1.8 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 1.2 ### 3.3 4.8 9.0 4.8
### ### ### 1.6 ### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### 1.1 ### 2.6 ### ### ###
### 1.1 1.2 1.4 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 2.1 1.8 1.6 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 1.4 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 1.1 1.2 ###
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should emit -5 signal with 5 max distance and 0.5 diffusion factor 1`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### -5.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### -3.5 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### -2.9 -2.5 -2.2 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ###
0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should emit 1 signal with 5 max distance and 0.5 diffusion factor 1`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 1.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 0.7 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.6 0.5 0.4 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ###
0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should emit 10 signal with 10 max distance and 0.9 diffusion factor 1`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 1.1 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 1.2 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ###
0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 1.1 1.2 1.4 1.6 1.8 2.1 2.6 ### 1.8 1.6 1.4 1.2 1.1 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 1.2 ### 2.1 ### 3.3 ### 2.1 ### 1.6 ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 1.1 0.0 1.1 ### 2.6 3.3 4.8 3.3 2.6 2.1 1.8 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 1.2 ### ### ### ### ### 9.0 ### 2.1 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 1.1 1.2 1.4 1.6 1.8 2.1 2.6 3.3 4.8 ### 1.8 1.6 1.4 1.2 1.1 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 1.2 ### 1.6 ### ### ### 3.3 ### 1.6 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 1.1 ### 1.4 1.2 1.1 ### 2.6 ### 1.4 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 1.2 ### 1.2 ### 2.1 ### 1.2 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 1.1 ### 1.4 1.6 1.8 1.6 1.4 1.2 1.1 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;
